[YUNIKORN-3109] Add comprehensive Autoscaler test suite

rrajesh-cloudera opened the pull request:
- Create new autoscaler test directory with 9 comprehensive test cases
- Add basic autoscaling functionality tests
- Add complex scenarios: preemption, multi-queue competition, gang scheduling
- Add proper lint compliance with nolint annotations for cleanup operations
- All tests pass with proper YuniKorn integration and resource management

Test coverage includes:
- Pod scaling with YuniKorn scheduler
- Resource utilization metrics for autoscaling
- Queue-based resource management
- Rapid scaling operations
- Preemption during autoscaling
- Multi-queue autoscaling with resource competition
- Autoscaling with preemption policies
- Rapid scaling with preemption cascades
- Gang scheduling autoscaling with preemption

### What is this PR for?
A few sentences describing the overall goals of the pull request's commits.
First time? Check out the contributing guide - http://yunikorn.apache.org/community/how_to_contribute   


### What type of PR is it?
* [ ] - Bug Fix
* [x] - Improvement
* [ ] - Feature
* [ ] - Documentation
* [ ] - Hot Fix
* [ ] - Refactoring

### Todos
* [ ] - Task

### What is the Jira issue?
https://issues.apache.org/jira/browse/YUNIKORN-3109

rrajesh-cloudera committed:
Add comprehensive autoscaler test suite with performance optimizations

- Create new autoscaler test directory with 9 comprehensive test cases
- Add basic autoscaling functionality tests
- Add complex scenarios: preemption, multi-queue competition, gang scheduling
- Optimize test performance: 47% faster execution (3m8s vs 5m54s)
- Reduce resource requests: CPU 200m->50m, Memory 256Mi->64Mi
- Optimize timeouts: 480s->90-150s with faster polling intervals
- Add proper lint compliance with nolint annotations for cleanup operations
- All tests pass with proper YuniKorn integration and resource management

Test coverage includes:
- Pod scaling with YuniKorn scheduler
- Resource utilization metrics for autoscaling
- Queue-based resource management
- Rapid scaling operations
- Preemption during autoscaling
- Multi-queue autoscaling with resource competition
- Autoscaling with preemption policies
- Rapid scaling with preemption cascades
- Gang scheduling autoscaling with preemption

rrajesh-cloudera commented:
@pbacsko @manirajv06

rrajesh-cloudera committed:
Fix flaky Verify_Rapid_Scaling_Operations test

- Replace WaitForPodBySelectorRunning with robust Eventually blocks to eliminate race conditions
- Add comprehensive debugging output for pod status tracking
- Wait for exact expected number of running pods instead of relying on flawed helper function
- Improve error handling and verification throughout the test

Fixes the core race condition where test would pass before all intended pods were actually running.

pbacsko commented on test/e2e/autoscaler/autoscaler_test.go:
```diff
@@ -0,0 +1,990 @@
+		defer func() { _ = kClient.DeleteDeployment(deploymentObj3.Name, ns) }() //nolint:errcheck // Cleanup operation
+
+		By("Verifying initial deployment state")
+		gomega.Eventually(func() bool {
+			podsInNs, getErr := kClient.GetPods(ns)
+			if getErr != nil {
+				return false
+			}
+			app1Running := 0
+			app2Running := 0
+			app3Running := 0
+
+			for _, pod := range podsInNs.Items {
+				if pod.Status.Phase == v1.PodRunning {
+					switch pod.Labels["app"] {
+					case app1Name:
+						app1Running++
+					case app2Name:
+						app2Running++
+					case app3Name:
+						app3Running++
+					}
+				}
+			}
+
+			return app1Running >= 2 && app2Running >= 2 && app3Running >= 2
+		}, 180*time.Second, 5*time.Second).Should(gomega.BeTrue())
```
Again, same behavior can be achieved with the new helper methods.
3 apps, 3 method calls:
kClient.WaitForPodCountBySelectorRunning(ns, "app=app1", 2, 60)
kClient.WaitForPodCountBySelectorRunning(ns, "app=app2", 2, 60)
kClient.WaitForPodCountBySelectorRunning(ns, "app=app3", 2, 60)

pbacsko commented on test/e2e/autoscaler/autoscaler_test.go:
```diff
@@ -0,0 +1,990 @@
+		}, 180*time.Second, 5*time.Second).Should(gomega.BeTrue())
+
+		By("Scaling all deployments up to trigger resource competition")
+		deployment1.Spec.Replicas = &[]int32{4}[0]
```
For easier readiability, just do this:
```
newReplicaCount = 4
deployment1.Spec.Replicas := &newReplicaCount
```

pbacsko commented on test/e2e/autoscaler/autoscaler_test.go:
```diff
@@ -0,0 +1,990 @@
+		Ω(err).NotTo(HaveOccurred())
+
+		By("Verifying resource competition and scaling behavior")
+		gomega.Eventually(func() bool {
+			podsInNs, getErr := kClient.GetPods(ns)
+			if getErr != nil {
+				return false
+			}
+			app1Running := 0
+			app2Running := 0
+			app3Running := 0
+			totalRunning := 0
+
+			for _, pod := range podsInNs.Items {
+				if pod.Status.Phase == v1.PodRunning {
+					totalRunning++
+					switch pod.Labels["app"] {
+					case app1Name:
+						app1Running++
+					case app2Name:
+						app2Running++
+					case app3Name:
+						app3Running++
+					}
+				}
+			}
+
+			return app1Running >= 3 && app2Running >= 3 && app3Running >= 2 && totalRunning >= 8
+		}, 90*time.Second, 3*time.Second).Should(gomega.BeTrue(), "All apps should scale up within resource constraints")
```
Same as above

pbacsko commented on test/e2e/autoscaler/autoscaler_test.go:
```diff
@@ -0,0 +1,990 @@
+		defer func() { _ = kClient.DeleteDeployment(allowDeploymentObj.Name, ns) }() //nolint:errcheck // Cleanup operation
+
+		By("Verifying initial deployment state")
+		gomega.Eventually(func() bool {
+			podsInNs, getErr := kClient.GetPods(ns)
+			if getErr != nil {
+				return false
+			}
+			fenceRunning := 0
+			allowRunning := 0
+
+			for _, pod := range podsInNs.Items {
+				if pod.Status.Phase == v1.PodRunning {
+					if strings.Contains(pod.Name, fenceApp) {
+						fenceRunning++
+					} else if strings.Contains(pod.Name, allowApp) {
+						allowRunning++
+					}
+				}
+			}
+
+			return fenceRunning >= 2 && allowRunning >= 2
+		}, 120*time.Second, 5*time.Second).Should(gomega.BeTrue())
```
Replace it with suggested helper methods.

pbacsko commented on test/e2e/autoscaler/autoscaler_test.go:
```diff
@@ -0,0 +1,990 @@
+		time.Sleep(30 * time.Second) // Allow scaling to start
+
+		// Create high-priority aggressor deployment
+		aggressorDeployment := createTestDeployment(aggressorApp, ns, 3, map[string]string{
```
Can't we call this `highPrioDeployment`? And in a similar vein, `highPrioApp`. The term "aggressor" sounds weird to me.

pbacsko commented on test/e2e/autoscaler/autoscaler_test.go:
```diff
@@ -0,0 +1,990 @@
+		defer func() { _ = kClient.DeleteDeployment(aggressorDeploymentObj.Name, ns) }() //nolint:errcheck // Cleanup operation
+
+		By("Verifying preemption behavior based on policies")
+		gomega.Eventually(func() bool {
+			podsInNs, getErr := kClient.GetPods(ns)
+			if getErr != nil {
+				return false
+			}
+
+			fenceRunning := 0
+			allowRunning := 0
+			aggressorRunning := 0
+
+			for _, pod := range podsInNs.Items {
+				if pod.Status.Phase == v1.PodRunning {
+					switch pod.Labels["app"] {
+					case fenceApp:
+						fenceRunning++
+					case allowApp:
+						allowRunning++
+					case aggressorApp:
+						aggressorRunning++
+					}
+				}
+			}
+
+			// Fence pods should maintain higher count (protected from preemption)
+			// Allow pods may be preempted
+			// Aggressor should get some resources
+			return aggressorRunning > 0 && fenceRunning >= allowRunning
+		}, 300*time.Second, 10*time.Second).Should(gomega.BeTrue())
```
Replace this with suggested helper methods.

pbacsko commented on test/e2e/autoscaler/autoscaler_test.go:
```diff
@@ -0,0 +1,990 @@
+
+		By("Waiting for all initial pods to be running")
+		for _, app := range applications {
+			gomega.Eventually(func() int {
+				podsInNs, getErr := kClient.GetPods(ns)
+				if getErr != nil {
+					return -1
+				}
+				runningCount := 0
+				for _, pod := range podsInNs.Items {
+					if strings.Contains(pod.Name, app.name) && pod.Status.Phase == v1.PodRunning {
+						runningCount++
+					}
+				}
+				return runningCount
```
Replace this with suggested helper method.

pbacsko commented on test/e2e/autoscaler/autoscaler_test.go:
```diff
@@ -0,0 +1,990 @@
+	"github.com/apache/yunikorn-k8shim/test/e2e/framework/helpers/yunikorn"
+)
+
+var _ = Describe("Autoscaler Tests", func() {
```
The word "Autoscaler" is misleading. That would imply that cluster autoscaling is involved, which is a completely different animal and that's not what we're testing in this suite.

I suggest calling it "ReplicaScaling" or sth similar.

pbacsko reviewed (changes requested):
Ok, here is the first round. My biggest complaint (similar to #987) is the overuse of `gomega.Eventually()` with sometimes very specific wait conditions which are hard to understand.
As a first step, let's try to reduce those. Use the simples possible condition - ideally an existing `Wait...()` method from `KubeCtl`.

Second, you can delete all defer cleanup calls.
Third, just use simple variable names and avoid tricky constructs like `&[]int32{4}[0]`.

rrajesh-cloudera committed:
Refactor autoscaler tests to replica scaling and fix race conditions

- Rename test directory: autoscaler/ → replica_scaling/
- Update all terminology from 'Autoscaler/Autoscaling' to 'ReplicaScaling'
- Fix misleading naming: tests focus on pod replica scaling, not cluster autoscaling
- Update package names, function names, test names, and documentation
- Fix race condition in WaitForPodBySelectorRunning helper function:
  * Now properly waits for pods to exist before checking running state
  * Uses polling to handle deployment controller timing
- All tests now pass: 9/9 success rate
- Consolidated helper functions and eliminated gomega.Eventually overuse
- Improved variable naming (aggressor → highPrio) for clarity

Fixes test failures in rapid scaling, preemption, and gang scheduling scenarios.

rrajesh-cloudera committed:
Fix race conditions in replica scaling tests

- Add WaitForPodCountBySelectorRunning helper function to wait for exact pod counts
- Fix rapid scaling test to properly wait for scale down to exactly 2 pods
- Improve timing reliability for CI environments
- Address assertion failure where expected 2 pods but got 3 during scale down
- All 9 replica scaling tests now pass consistently (100% success rate)

This resolves CI failures in the rapid scaling operations test.

pbacsko reviewed (changes requested):
I came up with a different: "Deployment Scaling" tests (so "deployment_scaling", "Deployment_Scaling", depending on where it occurs).
The tests could have a slightly more straightforward names, but let's get back to them in a later round.

Looking at the test scanarios, I beleve the following two are worth keeping:
- `Verify_Pod_Scaling_With_YuniKorn_Scheduler` --> scale 2->3 but also 3->2 (you can add extra step after 2->3 to see the new pod is indeed in the proper queue, etc)
- `Verify_Preemption_During_ReplicaScaling_Operations`

The rest of the tests are very complicated and difficult to understand. If they start to fail at any point during development, it might take hours or days to figure out what goes wrong. 

I definitely appreciate the time you put into writing them, but we need to look at these from different angles and the potential maintenance burden is significant.

rrajesh-cloudera committed:
Refactor replica scaling tests to deployment scaling with improved maintainability

- Renamed replica_scaling -> deployment_scaling for better terminology
- Simplified from 7+ complex tests to 2 core maintainable scenarios:
  * Verify_Pod_Scaling_With_YuniKorn_Scheduler (2->3->2 scaling with queue verification)
  * Verify_Preemption_During_Deployment_Scaling_Operations (priority-based preemption)
- Improved API server performance: polling interval 100ms -> 1s
- Enhanced code quality:
  * Consistent *v1.Pod pointer usage
  * Readable variable declarations (gracePeriod := int64(5))
  * Fixed linter warnings and improved type handling
- Made tests more robust with flexible YuniKorn allocation count checks
- All tests pass successfully with improved performance and maintainability

rrajesh-cloudera committed:
Address review comment: use time.Second instead of time.Second*2 in WaitForPodCountBySelectorRunning

This addresses the specific review comment about using consistent time.Second
interval for API polling in WaitForPodCountBySelectorRunning function,
changing from time.Second*2 to time.Second for better consistency.

rrajesh-cloudera committed:
Fix linter warning in WaitForSecret function

- Merge variable declaration with assignment (S1021)
- Changes: var cond wait.ConditionFunc; cond = func() to cond := wait.ConditionFunc(func()

rrajesh-cloudera committed:
updating polling time
